### DISCUSSION TITLE
run_tests error when compiling apps together

### DISCUSSION THREAD

**Discussion by aarograh** [idaholab/moose#21962] [Q&A Tools] [answered]
I'm trying to compile a couple of applicaitons together to perform some coupled calculations.  Everything compiles, but when I try to run tests, I get this error.  I'm not sure what's going on.  I don't get it with any of the standalone applications.

```
(moose) ag6@node20$  ./run_tests -j36 --dbg
Traceback (most recent call last):
  File "/home/ag6/zoo/mole/./run_tests", line 12, in <module>
    TestHarness.buildAndRun(sys.argv, 'mole', MOOSE_DIR)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/TestHarness.py", line 181, in buildAndRun
    harness.findAndRunTests()
  File "/home/ag6/zoo/mole/moose/python/TestHarness/TestHarness.py", line 418, in findAndRunTests
    self.scheduler.schedule(testers)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/schedulers/Scheduler.py", line 217, in schedule
    j_dag = jobs.createJobs(testers)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/JobDAG.py", line 56, in createJobs
    return self._checkDAG()
  File "/home/ag6/zoo/mole/moose/python/TestHarness/JobDAG.py", line 105, in _checkDAG
    self._doLast()
  File "/home/ag6/zoo/mole/moose/python/TestHarness/JobDAG.py", line 133, in _doLast
    self._doSkippedDependencies()
  File "/home/ag6/zoo/mole/moose/python/TestHarness/JobDAG.py", line 191, in _doSkippedDependencies
    if not job.getRunnable() or job.isFail() or job.isSkip():
  File "/home/ag6/zoo/mole/moose/python/TestHarness/schedulers/Job.py", line 153, in getRunnable
    return self.__tester.getRunnable(self.options)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/testers/Tester.py", line 238, in getRunnable
    self._runnable = self.checkRunnableBase(options)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/testers/Tester.py", line 597, in checkRunnableBase
    checks["registered_apps"] = util.getExeRegisteredApps(self.specs["executable"])
  File "/home/ag6/zoo/mole/moose/python/TestHarness/util.py", line 735, in getExeRegisteredApps
    data = getExeJSON(exe)
  File "/home/ag6/zoo/mole/moose/python/TestHarness/util.py", line 718, in getExeJSON
    output = output.split('**START JSON DATA**\n')[1]
IndexError: list index out of range
```

**Comment by milljm** [comment, marked as answer]
That is the application failing to produce proper JSON output. The TestHarness is asking the application for a dump of all your applications supported 'features'. Does the following not produce an error?

```bash
./your_application-opt --json
```

**Reply by aarograh** [reply]
Ah, yes, as it turns out that `--json` option does produce an error.  The output is below.  I am coupling the Mole application with Griffin.  The `AddSpeciesAction` is something inside Mole, so iot seems like perhaps some of the Mole components are not getting registered properly (interesting, because I didn't have this issue when coupling to a different application, but Griffin seems a lot more complicated in certain ways).  The output is below.  I'll start digging into it, but if you happen to have any quick suggestions please let me know and I'll try those.

```
Griffin version:         git commit 6ba94fd on 2022-08-29
Framework Information:
MOOSE Version:           git commit 16505c7 on 2022-03-28
LibMesh Version:
PETSc Version:           3.16.5
SLEPc Version:           3.16.2
Current Time:            Tue Aug 30 09:19:15 2022
Executable Timestamp:    Tue Aug 30 09:16:43 2022

Parallelism:
  Num Processors:        1
  Num Threads:           1

Command line: ./mole-dbg --json



*** ERROR ***
A 'AddSpeciesActionTest' is not a registered Action



Stack frames: 11
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
3: ActionFactory::getValidParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
4: Parser::buildJsonSyntaxTree(JsonSyntaxTree&) const
5: MooseApp::setupOptions()
6: RattlesnakeApp::setupOptions()
7: MooseApp::run()
8: main
9: __libc_start_main
10: ./mole-dbg(+0x3079) [0x555fda0bc079]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

**Reply by GiudGiud** [reply]
how is it registered inside Mole? at the top of the source file under the copyright header

**Reply by milljm** [reply]
Off topic, I see this error frequently on Discussions. I think I'll spend some time on the TestHarness to produce a better error message...

https://github.com/idaholab/moose/issues/21967

**Reply by aarograh** [reply]
@milljm that would be great!

@GiudGiud Here's the output of a source search for `AddSpeciesActionTest`.  I'm not sure which of these you're referring to, or perhaps I'm missing something completely?

```
mole/src/base/moleApp.C:
  63    registerSyntax("AddSpeciesAction", "Species/AddSpecies");
  64:   registerSyntax("AddSpeciesActionTest", "Species/AddSpeciesTest");
  65    registerSyntax("ConvectionAction", "Species/Convection");

mole/test/include/actions/AddSpeciesActionTest.h:
  4  
  5: class AddSpeciesActionTest : public AddSpeciesAction
  6  {

  8    static InputParameters validParams();
  9:   AddSpeciesActionTest(const InputParameters & params);
  10    virtual void act() override;

mole/test/src/actions/AddSpeciesActionTest.C:
   1: #include "AddSpeciesActionTest.h"
   2  #include "SpeciesDefinition.h"

   7  
   8: registerMooseAction("moleApp", AddSpeciesActionTest, "add_species_variables");
   9: registerMooseAction("moleApp", AddSpeciesActionTest, "add_species_definition");
  10: registerMooseAction("moleApp", AddSpeciesActionTest, "add_kernel");
  11: registerMooseAction("moleApp", AddSpeciesActionTest, "mole_test");
  12  
  13  InputParameters
  14: AddSpeciesActionTest::validParams()
  15  {

  21  
  22: AddSpeciesActionTest::AddSpeciesActionTest(const InputParameters & params)
  23    : AddSpeciesAction(params)

  27  void
  28: AddSpeciesActionTest::act()
  29  {
```

**Reply by GiudGiud** [reply]
so AddSpeciesActionsTest is only in the test directory, but is referred to in 
  64:   registerSyntax("AddSpeciesActionTest", "Species/AddSpeciesTest");
?

this seems problematic, unless Mole is always built with its test objects.

The problem is that nay combined application would not be built with all the sub-apps test objects
I think you ll want to add some #ifdefs in the source code to prevent that

**Reply by aarograh** [reply]
Hmm... ok... We had a lot of issues using the `moleTestApp` that was part of the original storked version of Mole.  It would work fine with mole standalone and mole coupled with SAM, but mole coupled with griffin would never work.  I kept getting a linking error that I was unable to resolve.  Perhaps that is related?  The specific linking error was that it could not find `moleTestApp::registerApps()` in `main.C` despite `registerApps` being clearly defined as part of the class in `moleTestApp.h` (which was also included in `main.C`).

All this registration stuff is pretty opaque to me.  Is there a guide somewhere on this stuff?  I think a better understanding of what's actually happening here would help solve a lot of challenges.  It's easy enough with just Mole but seems to get more complicated when other apps are involved.

**Reply by GiudGiud** [reply]
Dont think there is user documentation outside of the doxygen for it. 

moleTestApp will not be compiled by default for a combined app since it's a test 'object'.
Including the header is enough to let the compiler not error out because it sees the declaration, but the function definition will be missing.

Use make -n on the combined app to see which files are being compiled and check whether moleTestApp.C and AddSpeciesTestAction.C are being compiled

**Reply by aarograh** [reply]
Well I've removed `moleTestApp.C` completely from the repo so I know that's not being compiled.  I may back out that change and open a new discussion about that issue since I think it's separate from this one

**Reply by aarograh** [reply]
Ok... I found in another discussion that I was missing `BUILD_TEST_OBJECTS_LIB := yes` in my MakeFile (seems like that should have been there by default, but whatever), which is why test test app was having issues.  Switching back to the test app and then adding that line in the Makefile looks to have resolved all these issues.
